{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Capsotn project 1: Data input ,unify and cleanning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input genotypic and phenotypic data from soybase and the published paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Zixiang'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Data download from soybase websit: https://soybase.org/snps/index.php.\n",
    "df=pd.read_csv('soysnp50K_wm82.a2_41417.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           rs#                     ltName alleles chrom    pos PI86046  \\\n",
      "0  ss715578788  BARC_1.01_Gm_01_24939_A_G     A/G  GM01  24952       G   \n",
      "1  ss715578818  BARC_1.01_Gm_01_25990_C_T     C/T  GM01  26003       T   \n",
      "2  ss715578923  BARC_1.01_Gm_01_29658_A_G     A/G  GM01  29671       G   \n",
      "3  ss715578960  BARC_1.01_Gm_01_30699_G_A     G/A  GM01  30712       G   \n",
      "4  ss715579193  BARC_1.01_Gm_01_37005_C_T     C/T  GM01  37018       C   \n",
      "\n",
      "  PI90208 PI219698 PI253651A PI347550A   ...    PI574480B PI578360 PI578362  \\\n",
      "0       A        -         A         A   ...            A        A        G   \n",
      "1       C        -         C         C   ...            C        C        T   \n",
      "2       A        H         A         A   ...            A        A        G   \n",
      "3       G        G         G         G   ...            G        G        G   \n",
      "4       C        H         C         C   ...            C        C        T   \n",
      "\n",
      "  PI639693 PI657626 PI634759 PI423967 PI587906 PI587946 PI603516  \n",
      "0        A        A        A        A        A        A        H  \n",
      "1        C        C        C        C        C        C        H  \n",
      "2        A        A        A        A        A        A        H  \n",
      "3        G        G        G        G        G        G        G  \n",
      "4        C        C        C        C        C        C        -  \n",
      "\n",
      "[5 rows x 20092 columns]\n"
     ]
    }
   ],
   "source": [
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 42291 entries, 0 to 42290\n",
      "Columns: 20092 entries, rs# to PI603516\n",
      "dtypes: int64(1), object(20091)\n",
      "memory usage: 6.3+ GB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(df.info())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input my own genoeypic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2=pd.read_csv('own_g.hmp.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         rs# alleles  chrom    pos strand  assembly#  center  \\\n",
      "0  BARC_1.01_Gm_01_24939_A_G     A/G      1  24939      +        NaN     NaN   \n",
      "1  BARC_1.01_Gm_01_25990_C_T     C/T      1  25990      +        NaN     NaN   \n",
      "2  BARC_1.01_Gm_01_29658_A_G     A/G      1  29658      +        NaN     NaN   \n",
      "3  BARC_1.01_Gm_01_30699_G_A       G      1  30699      +        NaN     NaN   \n",
      "4  BARC_1.01_Gm_01_37005_C_T     C/T      1  37005      +        NaN     NaN   \n",
      "\n",
      "   protLSID  assayLSID  panelLSID   ...     2012-176 2012-177 2012-178  \\\n",
      "0       NaN        NaN        NaN   ...            G        A        G   \n",
      "1       NaN        NaN        NaN   ...            T        C        T   \n",
      "2       NaN        NaN        NaN   ...            G        A        G   \n",
      "3       NaN        NaN        NaN   ...            G        G        G   \n",
      "4       NaN        NaN        NaN   ...            T        C        T   \n",
      "\n",
      "  2012-180 2012-181 2012-182 2012-183 2012-184 2012-185 2012-186  \n",
      "0        A        A        A        A        A        A        A  \n",
      "1        C        C        C        C        C        C        C  \n",
      "2        A        A        A        A        A        A        A  \n",
      "3        G        G        G        G        G        G        G  \n",
      "4        C        C        C        C        C        C        C  \n",
      "\n",
      "[5 rows x 1238 columns]\n"
     ]
    }
   ],
   "source": [
    "print(df2.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Combine the public data with my own research data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Combined my own data with the big dataset\n",
    "combined_g = pd.merge(df,df2,left_on='ltName',right_on= 'rs#',how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         rs#_x                     ltName alleles_x chrom_x  pos_x PI86046  \\\n",
      "0  ss715578788  BARC_1.01_Gm_01_24939_A_G       A/G    GM01  24952       G   \n",
      "1  ss715578818  BARC_1.01_Gm_01_25990_C_T       C/T    GM01  26003       T   \n",
      "2  ss715578923  BARC_1.01_Gm_01_29658_A_G       A/G    GM01  29671       G   \n",
      "3  ss715578960  BARC_1.01_Gm_01_30699_G_A       G/A    GM01  30712       G   \n",
      "4  ss715579193  BARC_1.01_Gm_01_37005_C_T       C/T    GM01  37018       C   \n",
      "\n",
      "  PI90208 PI219698 PI253651A PI347550A   ...    2012-176 2012-177 2012-178  \\\n",
      "0       A        -         A         A   ...           G        A        G   \n",
      "1       C        -         C         C   ...           T        C        T   \n",
      "2       A        H         A         A   ...           G        A        G   \n",
      "3       G        G         G         G   ...           G        G        G   \n",
      "4       C        H         C         C   ...           T        C        T   \n",
      "\n",
      "  2012-180 2012-181 2012-182 2012-183 2012-184 2012-185 2012-186  \n",
      "0        A        A        A        A        A        A        A  \n",
      "1        C        C        C        C        C        C        C  \n",
      "2        A        A        A        A        A        A        A  \n",
      "3        G        G        G        G        G        G        G  \n",
      "4        C        C        C        C        C        C        C  \n",
      "\n",
      "[5 rows x 21330 columns]\n"
     ]
    }
   ],
   "source": [
    "print(combined_g.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "g_column_name=[cn for cn in combined_g.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rs#_x',\n",
       " 'ltName',\n",
       " 'alleles_x',\n",
       " 'chrom_x',\n",
       " 'pos_x',\n",
       " 'PI86046',\n",
       " 'PI90208',\n",
       " 'PI219698',\n",
       " 'PI253651A',\n",
       " 'PI347550A']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g_column_name[:10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Phenotype data downloaded from https://www.g3journal.org/content/9/7/2253.\n",
    "p_df = pd.read_csv('trait.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id   oil  protein  yield\n",
      "0  PI103414  21.6     37.7   2.28\n",
      "1  PI132215  20.4     39.7   1.84\n",
      "2  PI135589  15.5     39.1   1.91\n",
      "3  PI135590  16.0     41.0   1.61\n",
      "4  PI142491  19.8     39.9   1.82\n"
     ]
    }
   ],
   "source": [
    "print(p_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## appending my own phenotypic to the big data set\n",
    "p_own_df = pd.read_csv('p_own.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_py = p_df.append(p_own_df,ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Unify Genotypic and phenotypic data based on line's ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id   oil  protein  yield\n",
      "0  PI103414  21.6     37.7   2.28\n",
      "1  PI132215  20.4     39.7   1.84\n",
      "2  PI135589  15.5     39.1   1.91\n",
      "3  PI135590  16.0     41.0   1.61\n",
      "4  PI142491  19.8     39.9   1.82\n",
      "5  PI153214  18.2     41.1   1.82\n",
      "6  PI153229  19.5     38.3   1.74\n",
      "7  PI153236  20.4     39.0   2.07\n",
      "8  PI153247  20.0     39.5   1.95\n",
      "9  PI153255  20.8     36.9   2.20\n"
     ]
    }
   ],
   "source": [
    "print(combined_py[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_py.to_csv('combined_py.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_column_name=[item for item in combined_py.id]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PI103414', 'PI132215', 'PI135589', 'PI135590', 'PI142491', 'PI153214', 'PI153229', 'PI153236', 'PI153247', 'PI153255']\n"
     ]
    }
   ],
   "source": [
    "print(p_column_name[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "## find out common id among genotypic and phenotypic data\n",
    "common_name=[value for value in g_column_name if value in p_column_name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9821\n"
     ]
    }
   ],
   "source": [
    "print(len(common_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PI219698',\n",
       " 'PI253651A',\n",
       " 'PI347550A',\n",
       " 'PI408055A',\n",
       " 'PI408069',\n",
       " 'PI408169A',\n",
       " 'PI408169B',\n",
       " 'PI416923',\n",
       " 'PI417472A',\n",
       " 'PI417521']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_name[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_genotype=pd.concat([combined_g.loc[:, 'rs#_x':'pos_x'], combined_g[common_name]], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 42234 entries, 0 to 42233\n",
      "Columns: 9826 entries, rs#_x to 2012-186\n",
      "dtypes: int64(1), object(9825)\n",
      "memory usage: 3.1+ GB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(new_genotype.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         rs#_x                     ltName alleles_x chrom_x  pos_x PI219698  \\\n",
      "0  ss715578788  BARC_1.01_Gm_01_24939_A_G       A/G    GM01  24952        -   \n",
      "1  ss715578818  BARC_1.01_Gm_01_25990_C_T       C/T    GM01  26003        -   \n",
      "2  ss715578923  BARC_1.01_Gm_01_29658_A_G       A/G    GM01  29671        H   \n",
      "3  ss715578960  BARC_1.01_Gm_01_30699_G_A       G/A    GM01  30712        G   \n",
      "4  ss715579193  BARC_1.01_Gm_01_37005_C_T       C/T    GM01  37018        H   \n",
      "\n",
      "  PI253651A PI347550A PI408055A PI408069   ...    2012-176 2012-177 2012-178  \\\n",
      "0         A         A         A        A   ...           G        A        G   \n",
      "1         C         C         C        C   ...           T        C        T   \n",
      "2         A         A         A        A   ...           G        A        G   \n",
      "3         G         G         G        G   ...           G        G        G   \n",
      "4         C         C         C        C   ...           T        C        T   \n",
      "\n",
      "  2012-180 2012-181 2012-182 2012-183 2012-184 2012-185 2012-186  \n",
      "0        A        A        A        A        A        A        A  \n",
      "1        C        C        C        C        C        C        C  \n",
      "2        A        A        A        A        A        A        A  \n",
      "3        G        G        G        G        G        G        G  \n",
      "4        C        C        C        C        C        C        C  \n",
      "\n",
      "[5 rows x 9826 columns]\n"
     ]
    }
   ],
   "source": [
    "print(new_genotype.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_genotype.to_csv('new_genotyp_combined.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_pheno=combined_py.loc[combined_py['id'].isin(common_name)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_pheno.to_csv('new_phenotype_combined.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Cleaning for phenotypic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 12939 entries, 0 to 12938\n",
      "Data columns (total 5 columns):\n",
      "Unnamed: 0    12939 non-null int64\n",
      "id            12939 non-null object\n",
      "oil           12859 non-null float64\n",
      "protein       12859 non-null float64\n",
      "yield         10198 non-null float64\n",
      "dtypes: float64(3), int64(1), object(1)\n",
      "memory usage: 505.5+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "combined_p_non_dup = pd.read_csv('new_phenotype_combined.txt',sep=',')\n",
    "print(combined_p_non_dup.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Unnamed: 0        id   oil  protein  yield\n",
      "0           0  PI103414  21.6     37.7   2.28\n",
      "1           1  PI132215  20.4     39.7   1.84\n",
      "2           2  PI135589  15.5     39.1   1.91\n",
      "3           3  PI135590  16.0     41.0   1.61\n",
      "4           4  PI142491  19.8     39.9   1.82\n"
     ]
    }
   ],
   "source": [
    "## Note that there are duplicates in combined_p_non_dup\n",
    "print(combined_p_non_dup.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_p_non_dup=combined_p_non_dup.drop_duplicates(subset='id',keep='last')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 9821 entries, 0 to 12938\n",
      "Data columns (total 5 columns):\n",
      "Unnamed: 0    9821 non-null int64\n",
      "id            9821 non-null object\n",
      "oil           9754 non-null float64\n",
      "protein       9754 non-null float64\n",
      "yield         7099 non-null float64\n",
      "dtypes: float64(3), int64(1), object(1)\n",
      "memory usage: 460.4+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(combined_p_non_dup.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_p=combined_p_non_dup[['id','oil','protein','yield']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>oil</th>\n",
       "      <th>protein</th>\n",
       "      <th>yield</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>9754.000000</td>\n",
       "      <td>9754.000000</td>\n",
       "      <td>7099.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>18.492156</td>\n",
       "      <td>44.350415</td>\n",
       "      <td>2.215935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>2.148282</td>\n",
       "      <td>2.620695</td>\n",
       "      <td>0.994446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>8.200000</td>\n",
       "      <td>33.100000</td>\n",
       "      <td>0.030000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>17.200000</td>\n",
       "      <td>42.500000</td>\n",
       "      <td>1.510000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>18.700000</td>\n",
       "      <td>44.200000</td>\n",
       "      <td>2.080000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>20.000000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>2.890000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>25.600000</td>\n",
       "      <td>57.400000</td>\n",
       "      <td>5.384000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               oil      protein        yield\n",
       "count  9754.000000  9754.000000  7099.000000\n",
       "mean     18.492156    44.350415     2.215935\n",
       "std       2.148282     2.620695     0.994446\n",
       "min       8.200000    33.100000     0.030000\n",
       "25%      17.200000    42.500000     1.510000\n",
       "50%      18.700000    44.200000     2.080000\n",
       "75%      20.000000    46.000000     2.890000\n",
       "max      25.600000    57.400000     5.384000"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_p.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oil</th>\n",
       "      <th>protein</th>\n",
       "      <th>yield</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PI103414</td>\n",
       "      <td>21.6</td>\n",
       "      <td>37.7</td>\n",
       "      <td>2.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PI132215</td>\n",
       "      <td>20.4</td>\n",
       "      <td>39.7</td>\n",
       "      <td>1.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PI153214</td>\n",
       "      <td>18.2</td>\n",
       "      <td>41.1</td>\n",
       "      <td>1.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PI153229</td>\n",
       "      <td>19.5</td>\n",
       "      <td>38.3</td>\n",
       "      <td>1.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PI153236</td>\n",
       "      <td>20.4</td>\n",
       "      <td>39.0</td>\n",
       "      <td>2.07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id   oil  protein  yield\n",
       "0  PI103414  21.6     37.7   2.28\n",
       "1  PI132215  20.4     39.7   1.84\n",
       "5  PI153214  18.2     41.1   1.82\n",
       "6  PI153229  19.5     38.3   1.74\n",
       "7  PI153236  20.4     39.0   2.07"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_p.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id   oil  protein  yield  oil_1.5IQR  protein_1.5IQR  yield_1.5IQR\n",
      "0  PI103414  21.6     37.7   2.28       False           False         False\n",
      "1  PI132215  20.4     39.7   1.84       False           False         False\n",
      "5  PI153214  18.2     41.1   1.82       False           False         False\n",
      "6  PI153229  19.5     38.3   1.74       False           False         False\n",
      "7  PI153236  20.4     39.0   2.07       False           False         False\n"
     ]
    }
   ],
   "source": [
    "## To find out outliers using 1.5 IQR\n",
    "cols = list(combined_p.columns)\n",
    "cols.remove('id')\n",
    "combined_p[cols]\n",
    "## creat a for loop to add IQR juddgement for each trait\n",
    "for col in cols:\n",
    "    col_IQR = col + '_1.5IQR'\n",
    "    Q1 = combined_p[col].quantile(0.25)\n",
    "    Q3 = combined_p[col].quantile(0.75)\n",
    "    IQR = Q3 - Q1\n",
    "    combined_p[col_IQR] =(combined_p_non_dup[col]<(Q1 - 1.5 * IQR)) | (combined_p_non_dup[col] > (Q3 + 1.5 * IQR))\n",
    "\n",
    "print(combined_p.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 9815 entries, 0 to 12938\n",
      "Data columns (total 2 columns):\n",
      "id       9815 non-null object\n",
      "yield    7093 non-null float64\n",
      "dtypes: float64(1), object(1)\n",
      "memory usage: 230.0+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Construct a Boolean Series to identify outliers for yield trait\n",
    "yield_clean=combined_p[combined_p['yield_1.5IQR']==False][['id','yield']]\n",
    "print(yield_clean.info())\n",
    "yield_clean.to_csv('Yield.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 9680 entries, 0 to 12938\n",
      "Data columns (total 2 columns):\n",
      "id     9680 non-null object\n",
      "oil    9613 non-null float64\n",
      "dtypes: float64(1), object(1)\n",
      "memory usage: 226.9+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Construct a Boolean Series to identify outliers for oil trait\n",
    "oil_clean=combined_p[combined_p['oil_1.5IQR']==False][['id','oil']]\n",
    "print(oil_clean.info())\n",
    "oil_clean.to_csv('oil.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 9680 entries, 0 to 12938\n",
      "Data columns (total 2 columns):\n",
      "id     9680 non-null object\n",
      "oil    9613 non-null float64\n",
      "dtypes: float64(1), object(1)\n",
      "memory usage: 226.9+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Construct a Boolean Series to identify outliers for protein trait\n",
    "protein_clean=combined_p[combined_p['protein_1.5IQR']==False][['id','protein']]\n",
    "print(oil_clean.info())\n",
    "protein_clean.to_csv('protein.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Data Cleaning for genotypic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "import rpy2.robjects as robjects\n",
    "from rpy2.robjects.packages import importr\n",
    "import pandas as pd\n",
    "import argparse\n",
    "import sys\n",
    "import textwrap\n",
    "import timeit\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Backup for data cleaning quick start\n",
    "## For qucikly stating data cleaning\n",
    "### new_genotype = pd.read_csv('new_genotyp_combined.txt',sep=',')\n",
    "### combined_p_non_dup = pd.read_csv('new_phenotype_combined',sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_genotype = pd.read_csv('new_genotyp_combined.txt',sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "test=new_genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "test=pd.DataFrame(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "test.replace(to_replace='-', value=np.nan, inplace=True)\n",
    "test.replace(to_replace='N', value=np.nan, inplace=True)  \n",
    "test.replace(to_replace='W', value='H', inplace=True) \n",
    "test.replace(to_replace='S', value='H', inplace=True) \n",
    "test.replace(to_replace='M', value='H', inplace=True) \n",
    "test.replace(to_replace='K', value='H', inplace=True) \n",
    "test.replace(to_replace='R', value='H', inplace=True) \n",
    "test.replace(to_replace='Y', value='H', inplace=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>rs#_x</th>\n",
       "      <th>ltName</th>\n",
       "      <th>alleles_x</th>\n",
       "      <th>chrom_x</th>\n",
       "      <th>pos_x</th>\n",
       "      <th>PI219698</th>\n",
       "      <th>PI253651A</th>\n",
       "      <th>PI347550A</th>\n",
       "      <th>PI408055A</th>\n",
       "      <th>...</th>\n",
       "      <th>2012-176</th>\n",
       "      <th>2012-177</th>\n",
       "      <th>2012-178</th>\n",
       "      <th>2012-180</th>\n",
       "      <th>2012-181</th>\n",
       "      <th>2012-182</th>\n",
       "      <th>2012-183</th>\n",
       "      <th>2012-184</th>\n",
       "      <th>2012-185</th>\n",
       "      <th>2012-186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ss715578788</td>\n",
       "      <td>BARC_1.01_Gm_01_24939_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>24952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ss715578818</td>\n",
       "      <td>BARC_1.01_Gm_01_25990_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>26003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ss715578923</td>\n",
       "      <td>BARC_1.01_Gm_01_29658_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>29671</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ss715578960</td>\n",
       "      <td>BARC_1.01_Gm_01_30699_G_A</td>\n",
       "      <td>G/A</td>\n",
       "      <td>GM01</td>\n",
       "      <td>30712</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ss715579193</td>\n",
       "      <td>BARC_1.01_Gm_01_37005_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>37018</td>\n",
       "      <td>H</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9827 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0        rs#_x                     ltName alleles_x chrom_x  \\\n",
       "0           0  ss715578788  BARC_1.01_Gm_01_24939_A_G       A/G    GM01   \n",
       "1           1  ss715578818  BARC_1.01_Gm_01_25990_C_T       C/T    GM01   \n",
       "2           2  ss715578923  BARC_1.01_Gm_01_29658_A_G       A/G    GM01   \n",
       "3           3  ss715578960  BARC_1.01_Gm_01_30699_G_A       G/A    GM01   \n",
       "4           4  ss715579193  BARC_1.01_Gm_01_37005_C_T       C/T    GM01   \n",
       "\n",
       "   pos_x PI219698 PI253651A PI347550A PI408055A   ...    2012-176 2012-177  \\\n",
       "0  24952      NaN         A         A         A   ...           G        A   \n",
       "1  26003      NaN         C         C         C   ...           T        C   \n",
       "2  29671        H         A         A         A   ...           G        A   \n",
       "3  30712        G         G         G         G   ...           G        G   \n",
       "4  37018        H         C         C         C   ...           T        C   \n",
       "\n",
       "  2012-178 2012-180 2012-181 2012-182 2012-183 2012-184 2012-185 2012-186  \n",
       "0        G        A        A        A        A        A        A        A  \n",
       "1        T        C        C        C        C        C        C        C  \n",
       "2        G        A        A        A        A        A        A        A  \n",
       "3        G        G        G        G        G        G        G        G  \n",
       "4        T        C        C        C        C        C        C        C  \n",
       "\n",
       "[5 rows x 9827 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = list(test.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Since soybean is selfing species, it is important to find out accessions with high Heterozygous(H) genotype\n",
    "Column_Total=[]\n",
    "for col in cols:\n",
    "    count_h=0\n",
    "    for item in test[col]:\n",
    "        if item=='H':\n",
    "         count_h+=1\n",
    "    Column_Total.append(count_h/test.shape[0])\n",
    "test.loc['Column_H_Total']=Column_Total\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "## find out accessions with high proportion of missing value\n",
    "Column_missing=[]\n",
    "for col in cols:\n",
    "    count_m=0\n",
    "    for item in test[col]:\n",
    "        if str(item)=='nan':\n",
    "         count_m+=1\n",
    "    Column_missing.append(count_m/test.shape[0])\n",
    "test.loc['Column_Missing_rate']=Column_missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>rs#_x</th>\n",
       "      <th>ltName</th>\n",
       "      <th>alleles_x</th>\n",
       "      <th>chrom_x</th>\n",
       "      <th>pos_x</th>\n",
       "      <th>PI219698</th>\n",
       "      <th>PI253651A</th>\n",
       "      <th>PI347550A</th>\n",
       "      <th>PI408055A</th>\n",
       "      <th>...</th>\n",
       "      <th>2012-176</th>\n",
       "      <th>2012-177</th>\n",
       "      <th>2012-178</th>\n",
       "      <th>2012-180</th>\n",
       "      <th>2012-181</th>\n",
       "      <th>2012-182</th>\n",
       "      <th>2012-183</th>\n",
       "      <th>2012-184</th>\n",
       "      <th>2012-185</th>\n",
       "      <th>2012-186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42231</th>\n",
       "      <td>42231.0</td>\n",
       "      <td>ss715623959</td>\n",
       "      <td>BARC_1.01_Gm_16_28327950_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>SCAFFOLD_759</td>\n",
       "      <td>4974.0</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42232</th>\n",
       "      <td>42232.0</td>\n",
       "      <td>ss715584606</td>\n",
       "      <td>BARC_1.01_Gm_03_1776411_C_T</td>\n",
       "      <td>T/C</td>\n",
       "      <td>SCAFFOLD_843</td>\n",
       "      <td>3015.0</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42233</th>\n",
       "      <td>42233.0</td>\n",
       "      <td>ss715633065</td>\n",
       "      <td>BARC_1.01_Gm_19_11471731_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>SCAFFOLD_91</td>\n",
       "      <td>24874.0</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Column_H_Total</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.157243</td>\n",
       "      <td>0.000355164</td>\n",
       "      <td>0.0293839</td>\n",
       "      <td>0.000189421</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00641663</td>\n",
       "      <td>0.00584837</td>\n",
       "      <td>0.00182318</td>\n",
       "      <td>0.0116494</td>\n",
       "      <td>0.00272292</td>\n",
       "      <td>0.0140171</td>\n",
       "      <td>0.00485391</td>\n",
       "      <td>0.0106549</td>\n",
       "      <td>0.0224227</td>\n",
       "      <td>0.0564237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Column_Missing_rate</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0307328</td>\n",
       "      <td>0.00165739</td>\n",
       "      <td>0.0323428</td>\n",
       "      <td>0.000923405</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00535101</td>\n",
       "      <td>0.00561146</td>\n",
       "      <td>0.0049485</td>\n",
       "      <td>0.00883154</td>\n",
       "      <td>0.00509056</td>\n",
       "      <td>0.0128566</td>\n",
       "      <td>0.00487747</td>\n",
       "      <td>0.00755298</td>\n",
       "      <td>0.012904</td>\n",
       "      <td>0.0119332</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9827 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Unnamed: 0        rs#_x                        ltName  \\\n",
       "42231                   42231.0  ss715623959  BARC_1.01_Gm_16_28327950_C_T   \n",
       "42232                   42232.0  ss715584606   BARC_1.01_Gm_03_1776411_C_T   \n",
       "42233                   42233.0  ss715633065  BARC_1.01_Gm_19_11471731_C_T   \n",
       "Column_H_Total              0.0            0                             0   \n",
       "Column_Missing_rate         0.0            0                             0   \n",
       "\n",
       "                    alleles_x       chrom_x    pos_x   PI219698    PI253651A  \\\n",
       "42231                     C/T  SCAFFOLD_759   4974.0          C            C   \n",
       "42232                     T/C  SCAFFOLD_843   3015.0          T            T   \n",
       "42233                     C/T   SCAFFOLD_91  24874.0          C            C   \n",
       "Column_H_Total              0             0      0.0   0.157243  0.000355164   \n",
       "Column_Missing_rate         0             0      0.0  0.0307328   0.00165739   \n",
       "\n",
       "                     PI347550A    PI408055A    ...        2012-176  \\\n",
       "42231                        T            C    ...               T   \n",
       "42232                        T            T    ...               C   \n",
       "42233                        T            C    ...               C   \n",
       "Column_H_Total       0.0293839  0.000189421    ...      0.00641663   \n",
       "Column_Missing_rate  0.0323428  0.000923405    ...      0.00535101   \n",
       "\n",
       "                       2012-177    2012-178    2012-180    2012-181  \\\n",
       "42231                         C           C           C           C   \n",
       "42232                         C           C           C           C   \n",
       "42233                         T           C           C           C   \n",
       "Column_H_Total       0.00584837  0.00182318   0.0116494  0.00272292   \n",
       "Column_Missing_rate  0.00561146   0.0049485  0.00883154  0.00509056   \n",
       "\n",
       "                      2012-182    2012-183    2012-184   2012-185   2012-186  \n",
       "42231                        C           C           C        NaN          C  \n",
       "42232                        T           T           T          C          T  \n",
       "42233                        C           C           C          C          C  \n",
       "Column_H_Total       0.0140171  0.00485391   0.0106549  0.0224227  0.0564237  \n",
       "Column_Missing_rate  0.0128566  0.00487747  0.00755298   0.012904  0.0119332  \n",
       "\n",
       "[5 rows x 9827 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct a Boolean Series to identify outliers for genotypic data based on missing rate\n",
    "outliers_m_def= test.loc['Column_Missing_rate']<0.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_new=test.loc[:,outliers_m_def]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct a Boolean Series to identify outliers for genotypic data based on Hetozog rate\n",
    "outliers_h_def=test.loc['Column_H_Total']<0.10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct a Boolean Series to identify outliers for genotypic data based on Hetozog rate\n",
    "outliers_h_def=test.loc['Column_H_Total']<0.10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_new=test.loc[:,outliers_h_def]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>rs#_x</th>\n",
       "      <th>ltName</th>\n",
       "      <th>alleles_x</th>\n",
       "      <th>chrom_x</th>\n",
       "      <th>pos_x</th>\n",
       "      <th>PI253651A</th>\n",
       "      <th>PI347550A</th>\n",
       "      <th>PI408055A</th>\n",
       "      <th>PI408069</th>\n",
       "      <th>...</th>\n",
       "      <th>2012-176</th>\n",
       "      <th>2012-177</th>\n",
       "      <th>2012-178</th>\n",
       "      <th>2012-180</th>\n",
       "      <th>2012-181</th>\n",
       "      <th>2012-182</th>\n",
       "      <th>2012-183</th>\n",
       "      <th>2012-184</th>\n",
       "      <th>2012-185</th>\n",
       "      <th>2012-186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>ss715578788</td>\n",
       "      <td>BARC_1.01_Gm_01_24939_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>24952.0</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>ss715578818</td>\n",
       "      <td>BARC_1.01_Gm_01_25990_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>26003.0</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>ss715578923</td>\n",
       "      <td>BARC_1.01_Gm_01_29658_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>29671.0</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>ss715578960</td>\n",
       "      <td>BARC_1.01_Gm_01_30699_G_A</td>\n",
       "      <td>G/A</td>\n",
       "      <td>GM01</td>\n",
       "      <td>30712.0</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>ss715579193</td>\n",
       "      <td>BARC_1.01_Gm_01_37005_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>37018.0</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9771 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0        rs#_x                     ltName alleles_x chrom_x  \\\n",
       "0         0.0  ss715578788  BARC_1.01_Gm_01_24939_A_G       A/G    GM01   \n",
       "1         1.0  ss715578818  BARC_1.01_Gm_01_25990_C_T       C/T    GM01   \n",
       "2         2.0  ss715578923  BARC_1.01_Gm_01_29658_A_G       A/G    GM01   \n",
       "3         3.0  ss715578960  BARC_1.01_Gm_01_30699_G_A       G/A    GM01   \n",
       "4         4.0  ss715579193  BARC_1.01_Gm_01_37005_C_T       C/T    GM01   \n",
       "\n",
       "     pos_x PI253651A PI347550A PI408055A PI408069   ...    2012-176 2012-177  \\\n",
       "0  24952.0         A         A         A        A   ...           G        A   \n",
       "1  26003.0         C         C         C        C   ...           T        C   \n",
       "2  29671.0         A         A         A        A   ...           G        A   \n",
       "3  30712.0         G         G         G        G   ...           G        G   \n",
       "4  37018.0         C         C         C        C   ...           T        C   \n",
       "\n",
       "  2012-178 2012-180 2012-181 2012-182 2012-183 2012-184 2012-185 2012-186  \n",
       "0        G        A        A        A        A        A        A        A  \n",
       "1        T        C        C        C        C        C        C        C  \n",
       "2        G        A        A        A        A        A        A        A  \n",
       "3        G        G        G        G        G        G        G        G  \n",
       "4        T        C        C        C        C        C        C        C  \n",
       "\n",
       "[5 rows x 9771 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_last=test_new.iloc[:,6:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Minor Allele Frequency (MAF) refers to the frequency at which the least common allele occurs in a \n",
    "## given population. Low MAF SNPs tend to have poorly behaved test statistics(violation of large sample assumption). \n",
    "## The following function will help me finding out snps with low MAF and add a new column(maf) to the data set.\n",
    "\n",
    "def MAF(snps):\n",
    "    print()\n",
    "    counter = 0\n",
    "    \n",
    "    # Set up some constants for coding\n",
    "    major = '0'\n",
    "    minor = '2'\n",
    "    hetero = '1'\n",
    "    maf=[]      \n",
    "    for i in range(snps.shape[0]):\n",
    "        # Progress updates\n",
    "        counter += 1\n",
    "        if counter % 1e2 == 0:\n",
    "            print(\"Processed\", counter, \"of\", snps.shape[0] - 1, \"markers.\", end = '\\r')\n",
    "        \n",
    "        # Get allele counts\n",
    "        allele_counts = snps.iloc[i, :].value_counts()\n",
    "        major_allele, minor_allele, het_allele = \"N\", \"N\", \"N\"\n",
    "        major_count, minor_count, het_count = 0, 0, 0\n",
    "        \n",
    "        # Identify major and minor alleles\n",
    "        for k, v in allele_counts.iteritems():\n",
    "            if k in ['W', 'S', 'M', 'K', 'R', 'Y','H','0']:\n",
    "                het_allele, het_count = k, v\n",
    "            elif v > major_count:\n",
    "                minor_count, major_count = major_count, v\n",
    "                minor_allele, major_allele = major_allele, k\n",
    "            else:\n",
    "                minor_allele, minor_count = k, v\n",
    "        MAF_r=(minor_count/snps.shape[1] + (het_count/snps.shape[1])*0.5)\n",
    "        maf.append(MAF_r)\n",
    "    snps.loc[:,'maf'] = maf\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processed 42200 of 42235 markers.3000 of 42235 markers. 10100 of 42235 markers. markers.of 42235 markers. 42235 markers.markers. 32100 of 42235 markers. 35000 of 42235 markers.\r"
     ]
    }
   ],
   "source": [
    "MAF(test_last)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PI253651A</th>\n",
       "      <th>PI347550A</th>\n",
       "      <th>PI408055A</th>\n",
       "      <th>PI408069</th>\n",
       "      <th>PI408169A</th>\n",
       "      <th>PI408169B</th>\n",
       "      <th>PI416923</th>\n",
       "      <th>PI417472A</th>\n",
       "      <th>PI417521</th>\n",
       "      <th>PI423734</th>\n",
       "      <th>...</th>\n",
       "      <th>2012-177</th>\n",
       "      <th>2012-178</th>\n",
       "      <th>2012-180</th>\n",
       "      <th>2012-181</th>\n",
       "      <th>2012-182</th>\n",
       "      <th>2012-183</th>\n",
       "      <th>2012-184</th>\n",
       "      <th>2012-185</th>\n",
       "      <th>2012-186</th>\n",
       "      <th>maf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.243728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.242806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.256784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.082284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.152893</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9766 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  PI253651A PI347550A PI408055A PI408069 PI408169A PI408169B PI416923  \\\n",
       "0         A         A         A        A         G         A        A   \n",
       "1         C         C         C        C         T         C        C   \n",
       "2         A         A         A        A         G         A        A   \n",
       "3         G         G         G        G         G         G        G   \n",
       "4         C         C         C        C         T         C        C   \n",
       "\n",
       "  PI417472A PI417521 PI423734    ...    2012-177 2012-178 2012-180 2012-181  \\\n",
       "0         G        A        A    ...           A        G        A        A   \n",
       "1         T        C        C    ...           C        T        C        C   \n",
       "2         G        A        A    ...           A        G        A        A   \n",
       "3         A        G        G    ...           G        G        G        G   \n",
       "4         C        C        C    ...           C        T        C        C   \n",
       "\n",
       "  2012-182 2012-183 2012-184 2012-185 2012-186       maf  \n",
       "0        A        A        A        A        A  0.243728  \n",
       "1        C        C        C        C        C  0.242806  \n",
       "2        A        A        A        A        A  0.256784  \n",
       "3        G        G        G        G        G  0.082284  \n",
       "4        C        C        C        C        C  0.152893  \n",
       "\n",
       "[5 rows x 9766 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_last.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct a Boolean Series to identify outliers for genotypic data based on MAF(minor allele frequency) rate\n",
    "outliers_maf_def=test_last.maf>0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno_last=test_last.loc[outliers_maf_def]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PI253651A</th>\n",
       "      <th>PI347550A</th>\n",
       "      <th>PI408055A</th>\n",
       "      <th>PI408069</th>\n",
       "      <th>PI408169A</th>\n",
       "      <th>PI408169B</th>\n",
       "      <th>PI416923</th>\n",
       "      <th>PI417472A</th>\n",
       "      <th>PI417521</th>\n",
       "      <th>PI423734</th>\n",
       "      <th>...</th>\n",
       "      <th>2012-177</th>\n",
       "      <th>2012-178</th>\n",
       "      <th>2012-180</th>\n",
       "      <th>2012-181</th>\n",
       "      <th>2012-182</th>\n",
       "      <th>2012-183</th>\n",
       "      <th>2012-184</th>\n",
       "      <th>2012-185</th>\n",
       "      <th>2012-186</th>\n",
       "      <th>maf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.243728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.242806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.256784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.082284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.152893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.081669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.257962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.151869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.260010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.254173</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 9766 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  PI253651A PI347550A PI408055A PI408069 PI408169A PI408169B PI416923  \\\n",
       "0         A         A         A        A         G         A        A   \n",
       "1         C         C         C        C         T         C        C   \n",
       "2         A         A         A        A         G         A        A   \n",
       "3         G         G         G        G         G         G        G   \n",
       "4         C         C         C        C         T         C        C   \n",
       "5         C         C         C        C         C         C        C   \n",
       "6         A         A         A        A         G         A        A   \n",
       "7         G         G         G        G         A         G        G   \n",
       "8         C       NaN         C        C         T         C        C   \n",
       "9         C         C         C        C         A         C        C   \n",
       "\n",
       "  PI417472A PI417521 PI423734    ...    2012-177 2012-178 2012-180 2012-181  \\\n",
       "0         G        A        A    ...           A        G        A        A   \n",
       "1         T        C        C    ...           C        T        C        C   \n",
       "2         G        A        A    ...           A        G        A        A   \n",
       "3         A        G        G    ...           G        G        G        G   \n",
       "4         C        C        C    ...           C        T        C        C   \n",
       "5         A        C        C    ...           C        C        C        C   \n",
       "6         G        A        A    ...           A        G        A        A   \n",
       "7         G        G        G    ...           G        A        G        G   \n",
       "8         T        C        C    ...           C        T        C        C   \n",
       "9         A        C        C    ...           C        A        C        C   \n",
       "\n",
       "  2012-182 2012-183 2012-184 2012-185 2012-186       maf  \n",
       "0        A        A        A        A        A  0.243728  \n",
       "1        C        C        C        C        C  0.242806  \n",
       "2        A        A        A        A        A  0.256784  \n",
       "3        G        G        G        G        G  0.082284  \n",
       "4        C        C        C        C        C  0.152893  \n",
       "5        C        C        C        C        C  0.081669  \n",
       "6        A        A        A        A        A  0.257962  \n",
       "7        G        G        G        G        G  0.151869  \n",
       "8        C        C        C        C        C  0.260010  \n",
       "9        C        C        C        C        C  0.254173  \n",
       "\n",
       "[10 rows x 9766 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geno_last[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno_last.to_csv('geno_last.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Genotypic data Numericalization option 1 (Using Python function)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>rs#_x</th>\n",
       "      <th>ltName</th>\n",
       "      <th>alleles_x</th>\n",
       "      <th>chrom_x</th>\n",
       "      <th>pos_x</th>\n",
       "      <th>PI219698</th>\n",
       "      <th>PI253651A</th>\n",
       "      <th>PI347550A</th>\n",
       "      <th>PI408055A</th>\n",
       "      <th>...</th>\n",
       "      <th>2012-176</th>\n",
       "      <th>2012-177</th>\n",
       "      <th>2012-178</th>\n",
       "      <th>2012-180</th>\n",
       "      <th>2012-181</th>\n",
       "      <th>2012-182</th>\n",
       "      <th>2012-183</th>\n",
       "      <th>2012-184</th>\n",
       "      <th>2012-185</th>\n",
       "      <th>2012-186</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ss715578788</td>\n",
       "      <td>BARC_1.01_Gm_01_24939_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>24952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ss715578818</td>\n",
       "      <td>BARC_1.01_Gm_01_25990_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>26003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ss715578923</td>\n",
       "      <td>BARC_1.01_Gm_01_29658_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>29671</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ss715578960</td>\n",
       "      <td>BARC_1.01_Gm_01_30699_G_A</td>\n",
       "      <td>G/A</td>\n",
       "      <td>GM01</td>\n",
       "      <td>30712</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ss715579193</td>\n",
       "      <td>BARC_1.01_Gm_01_37005_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>37018</td>\n",
       "      <td>H</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 9827 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0        rs#_x                     ltName alleles_x chrom_x  \\\n",
       "0           0  ss715578788  BARC_1.01_Gm_01_24939_A_G       A/G    GM01   \n",
       "1           1  ss715578818  BARC_1.01_Gm_01_25990_C_T       C/T    GM01   \n",
       "2           2  ss715578923  BARC_1.01_Gm_01_29658_A_G       A/G    GM01   \n",
       "3           3  ss715578960  BARC_1.01_Gm_01_30699_G_A       G/A    GM01   \n",
       "4           4  ss715579193  BARC_1.01_Gm_01_37005_C_T       C/T    GM01   \n",
       "\n",
       "   pos_x PI219698 PI253651A PI347550A PI408055A   ...    2012-176 2012-177  \\\n",
       "0  24952      NaN         A         A         A   ...           G        A   \n",
       "1  26003      NaN         C         C         C   ...           T        C   \n",
       "2  29671        H         A         A         A   ...           G        A   \n",
       "3  30712        G         G         G         G   ...           G        G   \n",
       "4  37018        H         C         C         C   ...           T        C   \n",
       "\n",
       "  2012-178 2012-180 2012-181 2012-182 2012-183 2012-184 2012-185 2012-186  \n",
       "0        G        A        A        A        A        A        A        A  \n",
       "1        T        C        C        C        C        C        C        C  \n",
       "2        G        A        A        A        A        A        A        A  \n",
       "3        G        G        G        G        G        G        G        G  \n",
       "4        T        C        C        C        C        C        C        C  \n",
       "\n",
       "[5 rows x 9827 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## In this section, I'm going to compare two options (Python function and R code ) to numericalize the genotypic data, \n",
    "## and then compare the results.\n",
    "new_genotype.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "## The main dataset used \"H\" represent hetero genetype,\n",
    "## wheareas 'S','M', et. al were used to represent the H allele in my ow data. so unify is needed.\n",
    "df1=new_genotype.iloc[:,2:]\n",
    "df1.drop(['alleles_x', 'chrom_x', 'pos_x'],axis=1, inplace=True)\n",
    "df1=df1.T\n",
    "df1=df1.drop('ltName')\n",
    "df1.reindex()\n",
    "df1.replace(to_replace='-', value=np.nan, inplace=True)\n",
    "df1.replace(to_replace='N', value=np.nan, inplace=True)  \n",
    "df1.replace(to_replace='W', value='H', inplace=True) \n",
    "df1.replace(to_replace='S', value='H', inplace=True) \n",
    "df1.replace(to_replace='M', value='H', inplace=True) \n",
    "df1.replace(to_replace='K', value='H', inplace=True) \n",
    "df1.replace(to_replace='R', value='H', inplace=True) \n",
    "df1.replace(to_replace='Y', value='H', inplace=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_csv('sample_check.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "### The following readhmp function can be used in future.\n",
    "def readhmp(filename):\n",
    "    d = pd.read_csv(filename)\n",
    "    df1=d.iloc[:,2:]\n",
    "    df1.drop(['alleles_x', 'chrom_x', 'pos_x'],axis=1, inplace=True)\n",
    "    df1=df1.T\n",
    "    df1=df1.drop('ltName')\n",
    "    df1.reindex()\n",
    "    df1.replace(to_replace='-', value=np.nan, inplace=True)\n",
    "    df1.replace(to_replace='N', value=np.nan, inplace=True)\n",
    "    df1.replace(to_replace='W', value='H', inplace=True) \n",
    "    df1.replace(to_replace='S', value='H', inplace=True) \n",
    "    df1.replace(to_replace='M', value='H', inplace=True) \n",
    "    df1.replace(to_replace='K', value='H', inplace=True) \n",
    "    df1.replace(to_replace='R', value='H', inplace=True) \n",
    "    df1.replace(to_replace='Y', value='H', inplace=True) \n",
    "    return df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Use numericalize function from SNPtool(https://github.com/amkusmec/snptools)\n",
    "###  to convert data to major = '0', minor = '2', hetero = '1'.\n",
    "def numericalize(snps):\n",
    "    print()\n",
    "    counter = 0\n",
    "    \n",
    "    # Set up some constants for coding\n",
    "    major = '0'\n",
    "    minor = '2'\n",
    "    hetero = '1'\n",
    "            \n",
    "    for i in range(snps.shape[1]):\n",
    "        # Progress updates\n",
    "        counter += 1\n",
    "        if counter % 1e2 == 0:\n",
    "            print(\"Processed\", counter, \"of\", snps.shape[1] - 1, \"markers.\", end = '\\r')\n",
    "        \n",
    "        # Get allele counts\n",
    "        allele_counts = snps.iloc[:, i].value_counts()\n",
    "        major_allele, minor_allele, het_allele = \"N\", \"N\", \"N\"\n",
    "        major_count, minor_count, het_count = 0, 0, 0\n",
    "        \n",
    "        # Identify major and minor alleles\n",
    "        for k, v in allele_counts.iteritems():\n",
    "            if k in ['W', 'S', 'M', 'K', 'R', 'Y','H','0']:\n",
    "                het_allele, het_count = k, v\n",
    "            elif v > major_count:\n",
    "                minor_count, major_count = major_count, v\n",
    "                minor_allele, major_allele = major_allele, k\n",
    "            else:\n",
    "                minor_allele, minor_count = k, v\n",
    "        \n",
    "        # Replace alleles with numerical values\n",
    "        snps.iloc[:, i].replace(to_replace = [major_allele, minor_allele, het_allele], value = [major, minor, hetero], inplace = True)\n",
    "        snps.iloc[:, i] = pd.to_numeric(snps.iloc[:, i])\n",
    "        snps.iloc[:, i] = snps.iloc[:, i].fillna(major)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample=df1.iloc[:5,:20000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processed 20000 of 19999 markers.\r"
     ]
    }
   ],
   "source": [
    "numericalize(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>19990</th>\n",
       "      <th>19991</th>\n",
       "      <th>19992</th>\n",
       "      <th>19993</th>\n",
       "      <th>19994</th>\n",
       "      <th>19995</th>\n",
       "      <th>19996</th>\n",
       "      <th>19997</th>\n",
       "      <th>19998</th>\n",
       "      <th>19999</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PI219698</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI253651A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI347550A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408055A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408069</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 20000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0     1      2      3      4      5      6      7     8      9      \\\n",
       "PI219698      0     0      1      0      1      0      1      1     1      1   \n",
       "PI253651A     0     0      0      0      0      0      0      0     0      0   \n",
       "PI347550A     0     0      0      0      0      0      0      0     0      0   \n",
       "PI408055A     0     0      0      0      0      0      0      0     0      0   \n",
       "PI408069      0     0      0      0      0      0      0      0     0      0   \n",
       "\n",
       "           ...    19990  19991  19992  19993  19994  19995  19996  19997  \\\n",
       "PI219698   ...        0      0      0      0      0      0      0      0   \n",
       "PI253651A  ...        0      0      0      0      0      0      0      0   \n",
       "PI347550A  ...        0      0      2      0      0      0      0      0   \n",
       "PI408055A  ...        0      0      0      0      0      0      0      0   \n",
       "PI408069   ...        0      0      0      0      0      0      0      0   \n",
       "\n",
       "           19998  19999  \n",
       "PI219698       0      0  \n",
       "PI253651A      0      2  \n",
       "PI347550A      2      2  \n",
       "PI408055A      0      0  \n",
       "PI408069       0      0  \n",
       "\n",
       "[5 rows x 20000 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_csv('021format_of_g_combined.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotypic data Numericalization option 2 (runing R code in python)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Using R code to \n",
    "df2=new_genotype.iloc[:,2:]\n",
    "df2.replace(to_replace='-', value='N', inplace=True)\n",
    "df2.replace(to_replace='np.nan', value='N', inplace=True)\n",
    "df2.replace(to_replace='U', value='N', inplace=True)\n",
    "df2.replace(to_replace='W', value='H', inplace=True) \n",
    "df2.replace(to_replace='S', value='H', inplace=True) \n",
    "df2.replace(to_replace='M', value='H', inplace=True) \n",
    "df2.replace(to_replace='K', value='H', inplace=True) \n",
    "df2.replace(to_replace='R', value='H', inplace=True) \n",
    "df2.replace(to_replace='Y', value='H', inplace=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.iloc[0,4]='N'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.iloc[1,4]='N'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv('g_combined_last_s.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "## I connected the GAPIT.Numericalization and GAPIT.HapMap function. These R codes will help convert the hapmap data to (-1,0,1) data format\n",
    "rstring=\"\"\"\n",
    "GAPIT.Numericalization <-\n",
    "  function(x,bit=2,effect=\"Add\",impute=\"Major\", Create.indicator = FALSE, Major.allele.zero = TRUE, byRow=TRUE){\n",
    "    #Object: To convert character SNP genotpe to numerical\n",
    "    #Output: Coresponding numerical value\n",
    "    #Authors: Feng Tian and Zhiwu Zhang\n",
    "    # Last update: May 30, 2011 \n",
    "    ##############################################################################################\n",
    "    if(bit==1)  {\n",
    "      x[x==\"X\"]=\"N\"\n",
    "      x[x==\"-\"]=\"N\"\n",
    "      x[x==\"+\"]=\"N\"\n",
    "      x[x==\"/\"]=\"N\"\n",
    "      x[x==\"K\"]=\"Z\" #K (for GT genotype)is replaced by Z to ensure heterozygose has the largest value\n",
    "    }\n",
    "    \n",
    "    if(bit==2)  {\n",
    "      x[x==\"XX\"]=\"N\"\n",
    "      x[x==\"--\"]=\"N\"\n",
    "      x[x==\"++\"]=\"N\"\n",
    "      x[x==\"//\"]=\"N\"\n",
    "      x[x==\"NN\"]=\"N\"\n",
    "    }\n",
    "    \n",
    "    n=length(x)\n",
    "    lev=levels(as.factor(x))\n",
    "    lev=setdiff(lev,\"N\")\n",
    "    #print(lev)\n",
    "    len=length(lev)\n",
    "    #print(lev)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #Genotype counts\n",
    "    count=1:len\n",
    "    for(i in 1:len){\n",
    "      count[i]=length(x[(x==lev[i])])\n",
    "    }\n",
    "    \n",
    "    \n",
    "    \n",
    "    if(Major.allele.zero){\n",
    "      if(len>1 & len<=3){\n",
    "        #One bit: Make sure that the SNP with the major allele is on the top, and the SNP with the minor allele is on the second position\n",
    "        if(bit==1){ \n",
    "          count.temp = cbind(count, seq(1:len))\n",
    "          if(len==3) count.temp = count.temp[-3,]\n",
    "          count.temp <- count.temp[order(count.temp[,1], decreasing = TRUE),]\n",
    "          if(len==3)order =  c(count.temp[,2],3)else order = count.temp[,2]\n",
    "        }\n",
    "        \n",
    "        #Two bit: Make sure that the SNP with the major allele is on the top, and the SNP with the minor allele is on the third position\n",
    "        if(bit==2){ \n",
    "          count.temp = cbind(count, seq(1:len))\n",
    "          if(len==3) count.temp = count.temp[-2,]\n",
    "          count.temp <- count.temp[order(count.temp[,1], decreasing = TRUE),]\n",
    "          if(len==3) order =  c(count.temp[1,2],2,count.temp[2,2])else order = count.temp[,2]\n",
    "        }\n",
    "        \n",
    "        count = count[order]\n",
    "        lev = lev[order]\n",
    "        \n",
    "      }   #End  if(len<=1 | len> 3)\n",
    "    } #End  if(Major.allele.zero)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #make two  bit order genotype as AA,AT and TT, one bit as A(AA),T(TT) and X(AT)\n",
    "    if(bit==1 & len==3){\n",
    "      temp=count[2]\n",
    "      count[2]=count[3]\n",
    "      count[3]=temp\n",
    "    }\n",
    "    position=order(count)\n",
    "    \n",
    "    \n",
    "    #1status other than 2 or 3\n",
    "    if(len<=1 | len> 3)x=0\n",
    "    \n",
    "    #2 status\n",
    "    if(len==2)x=ifelse(x==\"N\",NA,ifelse(x==lev[1],0,2))\n",
    "    \n",
    "    #3 status\n",
    "    if(bit==1){\n",
    "      if(len==3)x=ifelse(x==\"N\",NA,ifelse(x==lev[1],0,ifelse(x==lev[3],1,2)))\n",
    "    }else{\n",
    "      if(len==3)x=ifelse(x==\"N\",NA,ifelse(x==lev[1],0,ifelse(x==lev[3],2,1)))\n",
    "    }\n",
    "    \n",
    "    #print(paste(lev,len,sep=\" \"))\n",
    "    #print(position)\n",
    "    \n",
    "    #missing data imputation\n",
    "    if(impute==\"Middle\") {x[is.na(x)]=1 }\n",
    "    \n",
    "    if(len==3){\n",
    "      if(impute==\"Minor\")  {x[is.na(x)]=position[1]  -1}\n",
    "      if(impute==\"Major\")  {x[is.na(x)]=position[len]-1}\n",
    "      \n",
    "    }else{\n",
    "      if(impute==\"Minor\")  {x[is.na(x)]=2*(position[1]  -1)}\n",
    "      if(impute==\"Major\")  {x[is.na(x)]=2*(position[len]-1)}\n",
    "    }\n",
    "    \n",
    "    #alternative genetic models\n",
    "    if(effect==\"Dom\") x=ifelse(x==1,1,0)\n",
    "    if(effect==\"Left\") x[x==1]=0\n",
    "    if(effect==\"Right\") x[x==1]=2\n",
    "    \n",
    "    if(byRow) {\n",
    "      result=matrix(x,n,1)\n",
    "    }else{\n",
    "      result=matrix(x,1,n)  \n",
    "    }\n",
    "    \n",
    "    return(result)\n",
    "  }#end of GAPIT.Numericalization function\n",
    "\n",
    "# Beginning of GAPIT.HapMap function\n",
    "\n",
    "\n",
    "GAPIT.HapMap <-\n",
    "  function(G,SNP.effect=\"Add\",SNP.impute=\"Major\",heading=TRUE, Create.indicator = FALSE, Major.allele.zero = TRUE){\n",
    "    #Object: To convert character SNP genotpe to numerical\n",
    "    #Output: Coresponding numerical value\n",
    "    #Authors: Feng Tian and Zhiwu Zhang\n",
    "    # Last update: May 30, 2011 \n",
    "    ##############################################################################################\n",
    "    \n",
    "    print(paste(\"Converting HapMap format to numerical under model of \", SNP.impute,sep=\"\"))\n",
    "    #gc()\n",
    "    #GAPIT.Memory.Object(name.of.trait=\"HapMap.Start\")\n",
    "    \n",
    "    #GT=data.frame(G[1,-(1:4)])\n",
    "    if(heading){\n",
    "      GT= t(G[1,-(1:4)])\n",
    "      GI= G[-1,c(1,3,4)]\n",
    "    }else{\n",
    "      GT=NULL\n",
    "      GI= G[,c(1,3,4)]\n",
    "    }\n",
    "    \n",
    "    \n",
    "    #Set column names\n",
    "    if(heading)colnames(GT)=\"taxa\"\n",
    "    colnames(GI)=c(\"SNP\",\"Chromosome\",\"Position\")\n",
    "    \n",
    "    #Initial GD\n",
    "    GD=NULL\n",
    "    bit=nchar(as.character(G[2,12])) #to determine number of bits of genotype\n",
    "    #print(paste(\"Number of bits for genotype: \", bit))\n",
    "    \n",
    "    print(\"Perform numericalization\")\n",
    "    \n",
    "    if(heading){\n",
    "      if(!Create.indicator) GD= apply(G[,-(1:5)],1,function(one) GAPIT.Numericalization(one,bit=bit,effect=SNP.effect,impute=SNP.impute, Major.allele.zero=Major.allele.zero))\n",
    "      if(Create.indicator) GD= t(G[-1,-(1:5)])\n",
    "    }else{\n",
    "      if(!Create.indicator) GD= apply(G[  ,-(1:5)],1,function(one) GAPIT.Numericalization(one,bit=bit,effect=SNP.effect,impute=SNP.impute, Major.allele.zero=Major.allele.zero))\n",
    "      if(Create.indicator) GD= t(G[ ,-(1:5)])\n",
    "    }\n",
    "    \n",
    "    #set GT and GI to NULL in case of null GD\n",
    "    if(is.null(GD)){\n",
    "      GT=NULL\n",
    "      GI=NULL\n",
    "    }\n",
    "    \n",
    "    \n",
    "    write.csv(GD, file =\"Numeric data (0,1,2).csv\")\n",
    "    print(\"The dimension of GD is:\")\n",
    "    print(dim(GD))\n",
    "    print (\"Succesfuly finished converting HapMap to numeric data !\")\n",
    "    \n",
    "    if(!Create.indicator) {print(paste(\"Succesfuly finished converting HapMap which has bits of \", bit,sep=\"\")) }\n",
    "    return(list(GT=GT,GD=GD,GI=GI))\n",
    "  }\n",
    "\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "readtable=robjects.r('read.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "transfunc=robjects.r(rstring)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_g_data=readtable(\"g_combined_last_s.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_df=transfunc(r_g_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_df_new=pd.read_csv('Numeric data (0,1,2).csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>V1</th>\n",
       "      <th>V2</th>\n",
       "      <th>V3</th>\n",
       "      <th>V4</th>\n",
       "      <th>V5</th>\n",
       "      <th>V6</th>\n",
       "      <th>V7</th>\n",
       "      <th>V8</th>\n",
       "      <th>V9</th>\n",
       "      <th>...</th>\n",
       "      <th>V42225</th>\n",
       "      <th>V42226</th>\n",
       "      <th>V42227</th>\n",
       "      <th>V42228</th>\n",
       "      <th>V42229</th>\n",
       "      <th>V42230</th>\n",
       "      <th>V42231</th>\n",
       "      <th>V42232</th>\n",
       "      <th>V42233</th>\n",
       "      <th>V42234</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42235 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  V1  V2  V3  V4  V5  V6  V7  V8  V9   ...    V42225  V42226  \\\n",
       "0           1   0   0   0   0   0   0   0   0   0   ...         0       0   \n",
       "1           2   0   0   0   0   0   0   0   0   0   ...         0       0   \n",
       "2           3   0   0   0   0   0   0   0   0   0   ...         0       0   \n",
       "3           4   0   0   0   0   0   0   0   0   0   ...         0       0   \n",
       "4           5   0   0   0   0   0   0   0   0   0   ...         0       0   \n",
       "\n",
       "   V42227  V42228  V42229  V42230  V42231  V42232  V42233  V42234  \n",
       "0       0       0       0       0       0       0       0       0  \n",
       "1       0       0       0       0       0       0       0       0  \n",
       "2       0       0       0       0       0       0       0       0  \n",
       "3       0       0       0       0       0       0       0       0  \n",
       "4       0       0       0       0       0       0       0       0  \n",
       "\n",
       "[5 rows x 42235 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_df_new.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_df_new.loc['Column_Total']= r_df_new.sum(numeric_only=True, axis=0)\n",
    "r_df_new.loc[:,'Row_Total'] = r_df_new.sum(numeric_only=True, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparing the output of the two numericalization options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=readhmp('new_sample_last.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 14 entries, PI219698 to PI424159A\n",
      "Columns: 42234 entries, 0 to 42233\n",
      "dtypes: object(42234)\n",
      "memory usage: 4.5+ MB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>42224</th>\n",
       "      <th>42225</th>\n",
       "      <th>42226</th>\n",
       "      <th>42227</th>\n",
       "      <th>42228</th>\n",
       "      <th>42229</th>\n",
       "      <th>42230</th>\n",
       "      <th>42231</th>\n",
       "      <th>42232</th>\n",
       "      <th>42233</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PI219698</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H</td>\n",
       "      <td>G</td>\n",
       "      <td>H</td>\n",
       "      <td>C</td>\n",
       "      <td>H</td>\n",
       "      <td>H</td>\n",
       "      <td>H</td>\n",
       "      <td>H</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI253651A</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI347550A</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408055A</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408069</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42234 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0     1     2     3     4     5     6     7     8     9      ...   \\\n",
       "PI219698    NaN   NaN     H     G     H     C     H     H     H     H  ...    \n",
       "PI253651A     A     C     A     G     C     C     A     G     C     C  ...    \n",
       "PI347550A     A     C     A     G     C     C     A     G   NaN     C  ...    \n",
       "PI408055A     A     C     A     G     C     C     A     G     C     C  ...    \n",
       "PI408069      A     C     A     G     C     C     A     G     C     C  ...    \n",
       "\n",
       "          42224 42225 42226 42227 42228 42229 42230 42231 42232 42233  \n",
       "PI219698      A     G     A     C     G     G     G     C     T     C  \n",
       "PI253651A     A     G     A     C     G     G     G     C     T     C  \n",
       "PI347550A     A     G     A     C     G     G     G     T     T     T  \n",
       "PI408055A     A     G     A     C     G     G     G     C     T     C  \n",
       "PI408069      A     G     A     C   NaN     G     G     C     T     C  \n",
       "\n",
       "[5 rows x 42234 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processed 42200 of 42233 markers.\r"
     ]
    }
   ],
   "source": [
    "numericalize(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PI219698</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI253651A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI347550A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408055A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408069</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408169A</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408169B</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI416923</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI417472A</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI417521</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI423734</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI423923</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI424145</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI424159A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           1  2  3  4\n",
       "PI219698   0  1  0  1\n",
       "PI253651A  0  0  0  0\n",
       "PI347550A  0  0  0  0\n",
       "PI408055A  0  0  0  0\n",
       "PI408069   0  0  0  0\n",
       "PI408169A  2  2  0  2\n",
       "PI408169B  0  0  0  0\n",
       "PI416923   0  0  0  0\n",
       "PI417472A  2  2  2  0\n",
       "PI417521   0  0  0  0\n",
       "PI423734   0  0  0  0\n",
       "PI423923   0  0  0  0\n",
       "PI424145   0  0  0  0\n",
       "PI424159A  0  0  0  0"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "inx=df.index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['PI219698', 'PI253651A', 'PI347550A', 'PI408055A', 'PI408069',\n",
       "       'PI408169A', 'PI408169B', 'PI416923', 'PI417472A', 'PI417521',\n",
       "       'PI423734', 'PI423923', 'PI424145', 'PI424159A'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2=pd.read_csv('new_sample_last.txt')\n",
    "df2=df2.iloc[:,2:]\n",
    "df2.replace(to_replace='-', value='N', inplace=True)\n",
    "df2.replace(to_replace='np.nan', value='N', inplace=True)\n",
    "df2.replace(to_replace='U', value='N', inplace=True)\n",
    "df2.replace(to_replace='W', value='H', inplace=True) \n",
    "df2.replace(to_replace='S', value='H', inplace=True) \n",
    "df2.replace(to_replace='M', value='H', inplace=True) \n",
    "df2.replace(to_replace='K', value='H', inplace=True) \n",
    "df2.replace(to_replace='R', value='H', inplace=True) \n",
    "df2.replace(to_replace='Y', value='H', inplace=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ltName</th>\n",
       "      <th>alleles_x</th>\n",
       "      <th>chrom_x</th>\n",
       "      <th>pos_x</th>\n",
       "      <th>PI219698</th>\n",
       "      <th>PI253651A</th>\n",
       "      <th>PI347550A</th>\n",
       "      <th>PI408055A</th>\n",
       "      <th>PI408069</th>\n",
       "      <th>PI408169A</th>\n",
       "      <th>PI408169B</th>\n",
       "      <th>PI416923</th>\n",
       "      <th>PI417472A</th>\n",
       "      <th>PI417521</th>\n",
       "      <th>PI423734</th>\n",
       "      <th>PI423923</th>\n",
       "      <th>PI424145</th>\n",
       "      <th>PI424159A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BARC_1.01_Gm_01_24939_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>24952</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BARC_1.01_Gm_01_25990_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>26003</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BARC_1.01_Gm_01_29658_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>29671</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BARC_1.01_Gm_01_30699_G_A</td>\n",
       "      <td>G/A</td>\n",
       "      <td>GM01</td>\n",
       "      <td>30712</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BARC_1.01_Gm_01_37005_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>37018</td>\n",
       "      <td>H</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BARC_1.01_Gm_01_38469_C_A</td>\n",
       "      <td>C/A</td>\n",
       "      <td>GM01</td>\n",
       "      <td>38482</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BARC_1.01_Gm_01_46360_A_G</td>\n",
       "      <td>A/G</td>\n",
       "      <td>GM01</td>\n",
       "      <td>46373</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BARC_1.01_Gm_01_49225_G_A</td>\n",
       "      <td>G/A</td>\n",
       "      <td>GM01</td>\n",
       "      <td>49238</td>\n",
       "      <td>H</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BARC_1.01_Gm_01_61178_C_T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>GM01</td>\n",
       "      <td>61191</td>\n",
       "      <td>H</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BARC_1.01_Gm_01_81983_C_A</td>\n",
       "      <td>C/A</td>\n",
       "      <td>GM01</td>\n",
       "      <td>81844</td>\n",
       "      <td>H</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ltName alleles_x chrom_x  pos_x PI219698 PI253651A  \\\n",
       "0  BARC_1.01_Gm_01_24939_A_G       A/G    GM01  24952        N         A   \n",
       "1  BARC_1.01_Gm_01_25990_C_T       C/T    GM01  26003        N         C   \n",
       "2  BARC_1.01_Gm_01_29658_A_G       A/G    GM01  29671        H         A   \n",
       "3  BARC_1.01_Gm_01_30699_G_A       G/A    GM01  30712        G         G   \n",
       "4  BARC_1.01_Gm_01_37005_C_T       C/T    GM01  37018        H         C   \n",
       "5  BARC_1.01_Gm_01_38469_C_A       C/A    GM01  38482        C         C   \n",
       "6  BARC_1.01_Gm_01_46360_A_G       A/G    GM01  46373        H         A   \n",
       "7  BARC_1.01_Gm_01_49225_G_A       G/A    GM01  49238        H         G   \n",
       "8  BARC_1.01_Gm_01_61178_C_T       C/T    GM01  61191        H         C   \n",
       "9  BARC_1.01_Gm_01_81983_C_A       C/A    GM01  81844        H         C   \n",
       "\n",
       "  PI347550A PI408055A PI408069 PI408169A PI408169B PI416923 PI417472A  \\\n",
       "0         A         A        A         G         A        A         G   \n",
       "1         C         C        C         T         C        C         T   \n",
       "2         A         A        A         G         A        A         G   \n",
       "3         G         G        G         G         G        G         A   \n",
       "4         C         C        C         T         C        C         C   \n",
       "5         C         C        C         C         C        C         A   \n",
       "6         A         A        A         G         A        A         G   \n",
       "7         G         G        G         A         G        G         G   \n",
       "8         N         C        C         T         C        C         T   \n",
       "9         C         C        C         A         C        C         A   \n",
       "\n",
       "  PI417521 PI423734 PI423923 PI424145 PI424159A  \n",
       "0        A        A        A        A         A  \n",
       "1        C        C        C        C         C  \n",
       "2        A        A        A        A         A  \n",
       "3        G        G        G        G         G  \n",
       "4        C        C        C        C         C  \n",
       "5        C        C        C        C         C  \n",
       "6        A        A        A        A         A  \n",
       "7        G        G        G        G         G  \n",
       "8        C        C        C        C         C  \n",
       "9        C        C        C        C         C  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv('g_sample.txt')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "readtable=robjects.r('read.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_g_data=readtable(\"g_sample.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "transfunc=robjects.r(rstring)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_df=transfunc(r_g_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_df_new=pd.read_csv('Numeric data (0,1,2).csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_df_new.index=df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>V1</th>\n",
       "      <th>V2</th>\n",
       "      <th>V3</th>\n",
       "      <th>V4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PI219698</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI253651A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI347550A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408055A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408069</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408169A</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI408169B</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI416923</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI417472A</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI417521</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI423734</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI423923</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI424145</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PI424159A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           V1  V2  V3  V4\n",
       "PI219698    0   0   1   0\n",
       "PI253651A   0   0   0   0\n",
       "PI347550A   0   0   0   0\n",
       "PI408055A   0   0   0   0\n",
       "PI408069    0   0   0   0\n",
       "PI408169A   2   2   2   0\n",
       "PI408169B   0   0   0   0\n",
       "PI416923    0   0   0   0\n",
       "PI417472A   2   2   2   2\n",
       "PI417521    0   0   0   0\n",
       "PI423734    0   0   0   0\n",
       "PI423923    0   0   0   0\n",
       "PI424145    0   0   0   0\n",
       "PI424159A   0   0   0   0"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_df_new.iloc[:,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_name=['id','py_t_1','py_t_2','py_t_3','R_t_1','R_t_2','R_t_3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "comp_df=pd.merge(r_df_new.iloc[:,1:4],df.iloc[:,:3],on=df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "comp_df.columns=col_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>py_t_1</th>\n",
       "      <th>py_t_2</th>\n",
       "      <th>py_t_3</th>\n",
       "      <th>R_t_1</th>\n",
       "      <th>R_t_2</th>\n",
       "      <th>R_t_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PI219698</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PI253651A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PI347550A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PI408055A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PI408069</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PI408169A</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PI408169B</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PI416923</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PI417472A</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PI417521</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PI423734</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PI423923</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PI424145</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PI424159A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  py_t_1  py_t_2  py_t_3 R_t_1 R_t_2  R_t_3\n",
       "0    PI219698       0       0       1     0     0      1\n",
       "1   PI253651A       0       0       0     0     0      0\n",
       "2   PI347550A       0       0       0     0     0      0\n",
       "3   PI408055A       0       0       0     0     0      0\n",
       "4    PI408069       0       0       0     0     0      0\n",
       "5   PI408169A       2       2       2     2     2      2\n",
       "6   PI408169B       0       0       0     0     0      0\n",
       "7    PI416923       0       0       0     0     0      0\n",
       "8   PI417472A       2       2       2     2     2      2\n",
       "9    PI417521       0       0       0     0     0      0\n",
       "10   PI423734       0       0       0     0     0      0\n",
       "11   PI423923       0       0       0     0     0      0\n",
       "12   PI424145       0       0       0     0     0      0\n",
       "13  PI424159A       0       0       0     0     0      0"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     False\n",
       "1      True\n",
       "2      True\n",
       "3      True\n",
       "4      True\n",
       "5      True\n",
       "6      True\n",
       "7      True\n",
       "8      True\n",
       "9      True\n",
       "10     True\n",
       "11     True\n",
       "12     True\n",
       "13     True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_df.py_t_1==comp_df.R_t_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     False\n",
       "1      True\n",
       "2      True\n",
       "3      True\n",
       "4      True\n",
       "5      True\n",
       "6      True\n",
       "7      True\n",
       "8      True\n",
       "9      True\n",
       "10     True\n",
       "11     True\n",
       "12     True\n",
       "13     True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_df.py_t_2==comp_df.R_t_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     True\n",
       "1     True\n",
       "2     True\n",
       "3     True\n",
       "4     True\n",
       "5     True\n",
       "6     True\n",
       "7     True\n",
       "8     True\n",
       "9     True\n",
       "10    True\n",
       "11    True\n",
       "12    True\n",
       "13    True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_df.py_t_3==comp_df.R_t_3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### Two numericalization options (python and R ) produced same results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
